### PR TITLE
Allow the sshkdf type to be passed as a single character

### LINKED ISSUE #14027: OSS-Fuzz issue #30037: Out-of-bounds read in kdf_sshkdf_set_ctx_params
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=30037

I pass the `OSSL_KDF_PARAM_SSHKDF_TYPE` parameter as a buffer of size 1: https://github.com/guidovranken/cryptofuzz/blob/f3bebc46e83f4a411cce95f9ca1cb5caadf146c9/modules/openssl/module.cpp#L2809-L2810

The buffer is not NULL-terminated because I'm specifying the size (1).

Unless I'm misunderstanding the documentation, the string does not have to be NULL-terminated if the size is specified:

```
OSSL_PARAM_construct_utf8_string() is a function that constructs a UTF8 string OSSL_PARAM structure.
A parameter with name key, storage buf and size bsize is created.
If bsize is zero, the string length is determined using strlen(3) + 1 for the null termination byte.
Generally pass zero for bsize instead of calling strlen(3) yourself.
```

(It seems a bit redundant to me to 1) specify the size and 2) have NULL-termination)

But `kdf_sshkdf_set_ctx_params` then treats it as a NULL-terminated string: https://github.com/openssl/openssl/blob/e947a0642db111bb34547b5f7d48e13163492ca5/providers/implementations/kdfs/sshkdf.c#L162

leading to out-of-bounds read.

### PR DESCRIPTION
**Pull request by levitte**
This partially reverts commit 270a5ce1d9ea579a2f1d45887971582b1ef2b6a1.

This also slightly modifies the way diverse parameters in are
specified in providers/fips/self_test_data.inc for better consistency.

Fixes #14027

### CODE REVIEW

**Comment by mattcaswell** [comment]
Looks ok, but CI failures appear relevant.

**Comment by levitte** [comment]
> Looks ok, but CI failures appear relevant.

It always worries me when gcc fails and clang succeeds or vice versa.  [sigh]

**Comment by t8m** [comment]
We should discuss this within OTC IMO.

**Comment by paulidale** [comment]
Agreed, we need to discussion the ramifications.

**Comment by levitte** [comment]
The CIs seem much happier, now that I fixed an "interesting" assumption in `self_test_kats.c`.

File: providers/implementations/kdfs/sshkdf.c
Comment (slontis) [inline providers/implementations/kdfs/sshkdf.c:166]: This seems like an incredibly awkward pattern.
Should char be its own type?

Comment (levitte) [reply providers/implementations/kdfs/sshkdf.c:166]: Ewwwwwwww...  so a separate OSSL_PARAM data type that *always* have data size 1?

**Comment by slontis** [comment]
Should this have a OTC hold on it?

File: providers/implementations/kdfs/sshkdf.c
Comment (J-Vin) [reply providers/implementations/kdfs/sshkdf.c:166]: 👍

Comment (levitte) [reply providers/implementations/kdfs/sshkdf.c:166]: With #14168, the data size is much more reliable, and this type of hackery won't be needed.

**Comment by levitte** [comment]
I've rebased this.  The SSHKDF get_params got significantly simpler after #14168 was merged

**Comment by slontis** [comment]
Errors are relevant.

**Comment by levitte** [comment]
> Errors are relevant.

Found the issue, it was a spot that should have been fixed in #14168, but wasn't

**Comment by levitte** [comment]
CIs are happy now

**Comment by t8m** [comment]
OTC: no objections to merging this

File: providers/fips/self_test_data.inc
Comment (t8m) [inline providers/fips/self_test_data.inc:20]: This is weird change, because ST_KAT_PARAM_UTF8STRING is used with C string literals elsewhere in the file. So better just not use it for the sshkdf_type case.

**Review by t8m — CHANGES_REQUESTED** [review]

File: providers/fips/self_test_data.inc
Comment (levitte) [reply providers/fips/self_test_data.inc:20]: So...  you'd rather I have it like this?

``` C
    { name, OSSL_PARAM_UTF8_STRING, &data, sizeof(data) }
```

Comment (levitte) [reply providers/fips/self_test_data.inc:20]: Ah, never mind, now I see what you're saying...

Comment (levitte) [reply providers/fips/self_test_data.inc:22]: I think I've fixed it

**Comment by t8m** [comment]
Now the CI fails and I suspect it is the crypto/params_from_text.c change.

**Comment by levitte** [comment]
> Now the CI fails and I suspect it is the crypto/params_from_text.c change.

Nope, it was the same kind of change *missing* in the OSSL_PARAM_BLD stuff

**Review by t8m — APPROVED** [review]
Approved if CI passes.

**Comment by levitte** [comment]
Merged

a8eb71ad577bbbd41cea915315451f0ef9f11581 Allow the sshkdf type to be passed as a single character
af8bd1d8359705c6a980c65b0c27c3e90fc43bea Fix OSSL_PARAM_allocate_from_text() for OSSL_PARAM_UTF8_STRING
6be27456e1346121b1fed797e92353733b59e16e Fix string termination and length setting in OSSL_PARAM_BLD_push_utf8_string()

**Pull request closed without merging**
